### PR TITLE
Pull snapshot data

### PR DESCRIPTION
**Pull request by motechFR**
Pull data from snapshot into the proposals page

### CODE REVIEW

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  const [proposal, setProposal] = useState<Proposal | null>(null);
+  const [snapshotProposal, setSnapshotProposal] = useState<SnapshotProposal | null>(null);
+  const [userVotes, setUserVotes] = useState<SnapshotVote[]>([]);
+  const [loading, setLoading] = useState(true);
+
+  const proposalEndDate = coerceToMilliseconds(snapshotProposal?.end ?? 0);
+
+  // Address to use for getting votes (95K aave)
+  const account = '0x070341aA5Ed571f0FB2c4a5641409B1A46b4961b';
```
Comment (mattcasey) [inline components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:33]: this should not be hard-coded?

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    charmClient.proposals.getProposal(proposalId)
+      .then(_proposal => setProposal(_proposal));
+
+  }, []);
+
+  const hasPassedDeadline = proposalEndDate < Date.now();
+
+  useEffect(() => {
+    if (snapshotProposal && hasPassedDeadline && proposal) {
```
Comment (mattcasey) [inline components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:62]: should put this in the backend

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  //  const account = '0x344d5A4da2329A885b24aeC4C8937e20Cd67b112';
+
+  // Either the number of votes or tokens
+
+  const voteChoices = snapshotProposal?.choices ?? [];
+  const voteScores = snapshotProposal?.scores ?? [];
+
+  useEffect(() => {
+
+    getSnapshotProposal(snapshotProposalId)
```
Comment (mattcasey) [inline components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:45]: try moving this to useSWR:
```
const { data: snapshotProposal } = useSWR('', () => getSnapshotProposal(snapshotProposalId));
```

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        });
+    }
+  }, [snapshotProposal, proposal]);
+  const remainingTime = relativeTime(proposalEndDate);
+
+  const currentUserChoices = userVotes.map(v => voteChoices[v.choice - 1]).join(',');
+
+  return (
+    <VotesWrapper id={`vote.${snapshotProposalId}`}>
+      <Box display='flex' justifyContent='space-between' alignItems='center' sx={{ mb: 2 }}>
```
Comment (mattcasey) [inline components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:75]: replace sx prop with `mb={2}`

File: components/common/CharmEditor/components/inlineVote/components/VotesWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import styled from '@emotion/styled';
+import FormControl from '@mui/material/FormControl';
+
+export const VotesWrapper = styled.div<{ detailed?: boolean }>`
+  background-color: ${({ theme, detailed }) => detailed && theme.palette.mode !== 'light' ? theme.palette.background.default : theme.palette.background.light};
+  padding: ${({ theme }) => theme.spacing(2)};
+  `;
+
+export const StyledFormControl = styled(FormControl)`
+border-bottom: 1px solid ${({ theme }) => theme.palette.divider};
```
Comment (mattcasey) [inline components/common/CharmEditor/components/inlineVote/components/VotesWrapper.tsx:10]: clean up whitespace?

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+  useEffect(() => {
+    if (snapshotProposal && hasPassedDeadline && proposal) {
+      charmClient.proposals.updateStatus(proposalId, 'vote_closed')
+        .then(updatedProposal => {
+          setProposal(updatedProposal);
+        });
+    }
+  }, [snapshotProposal, proposal]);
+  const remainingTime = relativeTime(proposalEndDate);
```
Comment (mattcasey) [inline components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:69]: we have a component RelativeTime which keeps the time up-to-date

**Review by mattcasey — APPROVED** [review]

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx (hunk 1 above)
Comment (motechFR) [reply components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx]: Done, was using for testing

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx (hunk 2 above)
Comment (motechFR) [reply components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx]: Our CRON should cover this

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx (hunk 3 above)
Comment (motechFR) [reply components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:37]: Done, will aim to use this pattern going forward

File: components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx (hunk 6 above)
Comment (motechFR) [reply components/common/CharmEditor/components/inlineVote/components/SnapshotVoteDetails.tsx:50]: Will leave as is for now. RelativeTime is more focused on seconds and minutes

**Pull request merged**